add flag to bypass jobspec validation

garlick opened the pull request:
This adds a FLUX_JOB_NOVALIDATE flag that can be passed by the instance owner to the `job-ingest` module to bypass jobspec validation on a per-job basis, as opposed to reloading the `job-ingest` module with the `disable-validator` module option and disabling the validator for all jobs.

I originally added this as part of the `flux job exec` work to improve turnaround time for short lived jobs.  I broke it out for separate consideration here, since it's not really all _that_ helpful for the case it was written for, as the cost of validation is low once the python co-process is up and running.

Can we imagine this being useful for other things?  I'm fine if we prefer to dump it.

grondo commented:
Seems like this could be useful when submitting a job that purposefully violates jobspec requirements, e.g. when testing Fluxion or another new scheduler, etc. Maybe even in the testsuite to ensure the job shell or other jobspec consumer can tolerate a bad jobspec (I think we already have some tests that "load a permissive validator")

However, you can also just reload the job-ingest module with `disable-validator`, so you'd have to have a situation where you want most jobspecs validated but not all of them...

garlick commented:
Oh yeah, `disable-validator` (corrected in description above).

Yeah, I'm not sure this is worth it.  Nix it?

grondo commented:
I don't know, it seems like it is pretty minimal and could be useful.. I say we merge this. Doesn't seem like it would get in the way at all :thinking:

garlick commented:
OK, forced a push with python format fix, and also added a tiny bit of test coverage.

garlick commented:
Added a bit more testing and rebased on current master.

garlick commented:
Coverage revealed that my test to have a guest try to use the flag wasn't resulting in coverage of the error handling.  Upon closer look, the test was failing at the command instead of in job-ingest.  Fixed that, then found a bug in the job-ingest error handling and fixed.

Go coverage tool!

grondo commented on src/bindings/python/flux/job/submit.py:
```diff
@@ -53,6 +54,10 @@ def submit_async(
         (default is False)
     :type pre_signed: bool
+    :param novalidate: jobspec does not need to be validated.
+        (default is False) Novalidate=True is restricted to the
```
```suggestion
        (default is False) novalidate=True is restricted to the
```

garlick replied:
I'll fix and set MWP if that's OK.

grondo reviewed (approved):
LGTM! One minor documentation typo that was hardly worth mentioning.
